### PR TITLE
Network Messages to Queue (Infra for Multi duties support) V2

### PR DESCRIPTION
**Pull request by nivBlox**

### CODE REVIEW

File: exporter/ibft/reader.go [hunk 1]
```diff
@@ -69,7 +69,7 @@ func (r *reader) Sync() error {
 		time.Sleep(1 * time.Second)
 	}()
 	netWaitGroup.Wait()
-	hs := ibftsync.NewHistorySync(r.logger, validatorPubkey, r.network, r.storage, r.validateDecidedMsg)
+	hs := ibftsync.NewHistorySync(r.logger, nil, validatorPubkey, r.network, r.storage, r.validateDecidedMsg) // TODO need to pass identifier
```
Comment (alonmuroch) [inline exporter/ibft/reader.go:72]: @nivBlox we should pass identifier and not pub key

File: ibft/sync/history.go [hunk 2]
```diff
@@ -20,19 +20,15 @@ type HistorySync struct {
...
-	validatorPK []byte,
-	network network.Network,
-	ibftStorage collections.Iibft,
-	validateDecidedMsgF func(msg *proto.SignedMessage) error,
-) *HistorySync {
+func NewHistorySync(logger *zap.Logger, identifier []byte, pubKey []byte, network network.Network, ibftStorage collections.Iibft, validateDecidedMsgF func(msg *proto.SignedMessage) error, ) *HistorySync {
 	return &HistorySync{
 		logger:              logger,
-		validatorPK:         validatorPK,
+		validatorPK:         pubKey,
```
Comment (alonmuroch) [inline ibft/sync/history.go:30]: @nivBlox we should drop validatorPK

File: ibft/sync/history_test.go [hunk 3]
```diff
@@ -158,7 +158,7 @@ func TestFetchDecided(t *testing.T) {
 			require.NoError(t, err)
 			storage := collections.NewIbft(db, logger, "attestation")
 			network := newTestNetwork(t, test.peers, int(test.rangeParams[2]), nil, test.decidedArr, nil)
-			s := NewHistorySync(logger, test.valdiatorPK, network, &storage, func(msg *proto.SignedMessage) error {
+			s := NewHistorySync(logger, test.valdiatorPK, test.valdiatorPK, network, &storage, func(msg *proto.SignedMessage) error {
```
Comment (alonmuroch) [inline ibft/sync/history_test.go:161]: @nivBlox should be identifier

File: ibft/sync/history_test.go [hunk 4]
```diff
@@ -305,13 +305,9 @@ func TestFindHighest(t *testing.T) {
...
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			s := NewHistorySync(zap.L(),
-				test.valdiatorPK,
-				newTestNetwork(t, test.peers, 100, test.highestMap, nil, nil),
-				nil,
-				func(msg *proto.SignedMessage) error {
-					return nil
-				})
+			s := NewHistorySync(zap.L(), test.valdiatorPK, test.valdiatorPK, newTestNetwork(t, test.peers, 100, test.highestMap, nil, nil), nil, func(msg *proto.SignedMessage) error {
```
Comment (alonmuroch) [inline ibft/sync/history_test.go:308]: @nivBlox should be identifier not validatorPK

File: ibft/utils.go [hunk 5]
```diff
@@ -0,0 +1,12 @@
+package ibft
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/bloxapp/ssv/beacon"
+)
+
+// IdentifierFormat return base format for lambda
+func IdentifierFormat(pubKey []byte, role beacon.Role) string {
```
Comment (alonmuroch) [inline ibft/utils.go:10]: @nivBlox why here and not in the validator package?

File: network/p2p/p2p_stream.go [hunk 6]
```diff
@@ -68,7 +68,11 @@ func (n *p2pNetwork) handleStream() {
 				switch cm.Type {
 				case network.NetworkMsg_SyncType:
 					cm.SyncMessage.FromPeerID = stream.Conn().RemotePeer().String()
-					ls.syncCh <- &network.SyncChanObj{Msg: cm.SyncMessage, Stream: netSyncStream}
+					ls.syncCh <- &network.SyncChanObj{
+						Lambda: cm.Lambda,
```
Comment (alonmuroch) [inline network/p2p/p2p_stream.go:72]: @nivBlox let's remove lambda from the SyncChanObj .. we have it in the sync msg

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: exporter/ibft/reader.go (hunk 1 above)
Comment (nivBlox) [reply exporter/ibft/reader.go:72]: @amir-blox will add the appropriate identifier

File: ibft/utils.go (hunk 5 above)
Comment (nivBlox) [reply ibft/utils.go:10]: identifier is for each ibft not validator so i think it should be under ibft

File: ibft/instance.go [hunk 7]
```diff
@@ -260,7 +259,7 @@ func (i *Instance) SignAndBroadcast(msg *proto.Message) error {
 		SignerIds: []uint64{i.ValidatorShare.NodeID},
 	}
 	if i.network != nil {
-		return i.network.Broadcast(signedMessage)
+		return i.network.Broadcast(i.ValidatorShare.PublicKey.Serialize(), signedMessage)
```
Comment (alonmuroch) [inline ibft/instance.go:262]: @nivBlox why we need to pass public key? not better to manage it with identifier?

File: ibft/sync/history.go [hunk 8]
```diff
@@ -16,23 +17,19 @@ import (
 // fetching decided messages from the network
 type HistorySync struct {
 	logger              *zap.Logger
+	publicKey           *bls.PublicKey
 	network             network.Network
 	ibftStorage         collections.Iibft
 	validateDecidedMsgF func(msg *proto.SignedMessage) error
-	validatorPK         []byte
```
Comment (alonmuroch) [inline ibft/sync/history.go:22]: @nivBlox why change name or have validatorPK at all?

File: ibft/utils.go (hunk 5 above)
Comment (alonmuroch) [reply ibft/utils.go:10]: @nivBlox yeah but it's not set in the iBFT package.. I'm looking at the code and it's actually not used at all.

File: network/network.go [hunk 9]
```diff
@@ -32,32 +32,32 @@ type SyncStream interface {
 // Network represents the behavior of the network
 type Network interface {
 	// Broadcast propagates a signed message to all peers
-	Broadcast(msg *proto.SignedMessage) error
+	Broadcast(topicName []byte, msg *proto.SignedMessage) error
```
Comment (alonmuroch) [inline network/network.go:35]: @nivBlox why did we add topicName?

File: storage/collections/ibft_storage.go [hunk 10]
```diff
@@ -14,17 +14,17 @@ import (
...
-	SaveCurrentInstance(pk []byte, state *proto.State) error
+	SaveCurrentInstance(identifier []byte, state *proto.State) error
 	// GetCurrentInstance returns the state for the current running (not yet decided) instance
-	GetCurrentInstance(pk []byte) (*proto.State, error)
+	GetCurrentInstance(identifier []byte) (*proto.State, error)
 	// SaveDecided saves a signed message for an ibft instance with decided justification
 	SaveDecided(signedMsg *proto.SignedMessage) error
 	// GetDecided returns a signed message for an ibft instance which decided by identifier
-	GetDecided(pk []byte, seqNumber uint64) (*proto.SignedMessage, error)
+	GetDecided(lambda []byte, seqNumber uint64) (*proto.SignedMessage, error)
```
Comment (alonmuroch) [inline storage/collections/ibft_storage.go:23]: @nivBlox let's set on a single name for this.. lambda or identifier.. I think identifier

File: validator/duty_execution.go [hunk 11]
```diff
@@ -88,10 +88,9 @@ func (v *Validator) postConsensusDutyExecution(ctx context.Context, logger *zap.
 
 	identifier := v.ibfts[role].GetIdentifier()
 	// TODO - should we construct it better?
-	if err := v.network.BroadcastSignature(&proto.SignedMessage{
+	if err := v.network.BroadcastSignature(v.Share.PublicKey.Serialize(), &proto.SignedMessage{
```
Comment (alonmuroch) [inline validator/duty_execution.go:91]: @nivBlox why not use identifier?

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: validator/duty_execution.go (hunk 11 above)
Comment (nivBlox) [reply validator/duty_execution.go:91]: cause the topics are pubkey based

File: storage/collections/ibft_storage.go (hunk 10 above)
Comment (nivBlox) [reply storage/collections/ibft_storage.go:23]: will change

File: network/network.go (hunk 9 above)
Comment (nivBlox) [reply network/network.go:35]: we used the pubkey from the msg but now that we removed it we need to pass it in order to fetch the topic

File: ibft/sync/history.go (hunk 8 above)
Comment (nivBlox) [reply ibft/sync/history.go:22]: for the get topic

File: ibft/instance.go (hunk 7 above)
Comment (nivBlox) [reply ibft/instance.go:262]: to get the topic

File: ibft/ibft_decided.go [hunk 12]
```diff
@@ -75,7 +73,7 @@ func (i *ibftImpl) ProcessDecidedMessage(msg *proto.SignedMessage) {
 			i.currentInstance.Stop()
 		}
 		// sync
-		s := ibft_sync.NewHistorySync(i.logger, msg.Message.ValidatorPk, i.network, i.ibftStorage, i.validateDecidedMsg)
+		s := ibft_sync.NewHistorySync(i.logger, i.ValidatorShare.PublicKey, i.GetIdentifier(), i.network, i.ibftStorage, i.validateDecidedMsg)
```
[the remark this answers is not in the document]
Comment (nivBlox) [reply ibft/ibft_decided.go:76]: to get the topic

File: exporter/ibft/reader.go [hunk 13]
```diff
@@ -69,7 +68,7 @@ func (r *reader) Sync() error {
 		time.Sleep(1 * time.Second)
 	}()
 	netWaitGroup.Wait()
-	hs := ibftsync.NewHistorySync(r.logger, validatorPubkey, r.network, r.storage, r.validateDecidedMsg)
+	hs := ibftsync.NewHistorySync(r.logger, r.validatorShare.PublicKey, nil, r.network, r.storage, r.validateDecidedMsg) // TODO need to pass identifier
```
[the remark this answers is not in the document]
Comment (nivBlox) [reply exporter/ibft/reader.go:71]: in order to get the tipic

File: exporter/ibft/reader.go (hunk 1 above)
Comment (alonmuroch) [reply exporter/ibft/reader.go]: but we are still passing validatorPK

File: exporter/ibft/reader.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alonmuroch) [reply exporter/ibft/reader.go]: mm I suggest we open an issue to handle it.. better have 1 place to handle mapping between identifier and pub keys than pass it around everywhere

File: exporter/ibft/reader.go (hunk 1 above)
Comment (nivBlox) [reply exporter/ibft/reader.go]: we need to pass the pubkey for the getTopic func

File: ibft/utils.go (hunk 5 above)
Comment (alonmuroch) [reply ibft/utils.go]: @nivBlox ?

File: exporter/ibft/reader.go (hunk 1 above)
Comment (alonmuroch) [reply exporter/ibft/reader.go]: @nivBlox please open an issue for it

Comment (nivBlox) [reply exporter/ibft/reader.go]: https://github.com/bloxapp/ssv/issues/138

**Pull request merged**